### PR TITLE
Propagate bounds with IA_MIP.

### PR DESCRIPTION
**Pull request by hongkai-dai**
First propagate the bounds by IA, and then tighten the bounds by MIP.
Note that the bounds gradient is incorrect if the bounds can't be active.

### CODE REVIEW

File: neural_network_lyapunov/relu_to_optimization.py [hunk 1]
```diff
@@ -561,6 +562,50 @@ def _compute_layer_bound(self,
...
+                            z_pre_relu_lo[neuron_index] = z_pre_relu_lo_ia[j]
+                        else:
+                            if z_pre_relu_lo[neuron_index] > 0:
+                                # If the MIP lower bound is > 0, then the ReLU
+                                # unit should always be active. We reduce this
+                                # MIP lower bound a little bit (but still
+                                # remains positive)
+                                z_pre_relu_lo[neuron_index] *= 0.99
+                                z_pre_relu_lo[
+                                    neuron_index] += 0.01 * torch.maximum(
```
Comment (blandry) [inline neural_network_lyapunov/relu_to_optimization.py:587]: maybe just use torch.clamp for this? avoids creating a tensor here (and having to go back to add .device on everything once we move to GPUs ;) )

File: neural_network_lyapunov/relu_to_optimization.py [hunk 2]
```diff
@@ -561,6 +562,50 @@ def _compute_layer_bound(self,
...
+                                z_pre_relu_lo[
+                                    neuron_index] += 0.01 * z_pre_relu_lo_ia[j]
+                        if torch.abs(z_pre_relu_up[neuron_index] -
+                                     z_pre_relu_up_ia[j]) < 1E-5:
+                            z_pre_relu_up[neuron_index] = z_pre_relu_up_ia[j]
+                        else:
+                            if z_pre_relu_up[neuron_index] <= 0:
+                                z_pre_relu_up[neuron_index] *= 0.99
+                                z_pre_relu_up[
+                                    neuron_index] += 0.01 * torch.minimum(
```
Comment (blandry) [inline neural_network_lyapunov/relu_to_optimization.py:601]: same here

File: neural_network_lyapunov/relu_to_optimization.py [hunk 3]
```diff
@@ -611,6 +663,20 @@ def _compute_network_output_bounds(self,
                             network_input_up.detach().numpy(),
                             create_prog_callback,
                             binary_var_type)
+                if method == mip_utils.PropagateBoundsMethod.IA_MIP:
+                    if torch.abs(linear_output_lo[i] -
+                                 linear_output_lo_ia[i]) < 1E-3:
```
Comment (blandry) [inline neural_network_lyapunov/relu_to_optimization.py:668]: why are we considering constraints active with such a higher threshold on the output layer?

File: neural_network_lyapunov/relu_to_optimization.py [hunk 4]
```diff
@@ -561,6 +562,50 @@ def _compute_layer_bound(self,
...
+                    # close to the MIP bound, then we use the IA bound.
+                    # Otherwise, we relax the MIP bound a little bit to make
+                    # sure this relaxed bound will not be active.
+                    z_pre_relu_lo_ia, z_pre_relu_up_ia = \
+                        mip_utils.compute_range_by_IA(
+                            self.model[2 * layer_count].weight, bias,
+                            linear_layer_input_lo, linear_layer_input_up)
+                    for j in range(self.model[2 * layer_count].out_features):
+                        neuron_index = self.relu_unit_index[layer_count][j]
+                        if torch.abs(z_pre_relu_lo[neuron_index] -
```
Comment (blandry) [inline neural_network_lyapunov/relu_to_optimization.py:576]: I'm definitely a little bit worried that this depends on the tolerance we use on the solver for the MIP bound. If we ever play with that to improve efficiency it might bite us. I think it's fine, but something to keep in mind.

**Review by blandry — APPROVED** [review]
I'm wondering if it might be worth zero'ing out the gradients on the case where the bounds have been relaxed. I guess the idea is that these bounds shouldn't end up active, and therefore won't affect the overall MILP gradient, but I think the current implementation would end up with the  gradient of those relaxed bounds corresponding to `.01 * ia_bound_grad` which seems like a weird gradient to keep around instead of the technically incorrect but more predictable `0`?

File: neural_network_lyapunov/relu_to_optimization.py (hunk 1 above)
Comment (hongkai-dai) [reply neural_network_lyapunov/relu_to_optimization.py:587]: Done. Good call.

File: neural_network_lyapunov/relu_to_optimization.py (hunk 4 above)
Comment (hongkai-dai) [reply neural_network_lyapunov/relu_to_optimization.py:576]: It is better to use a tolerance larger than the solver accuracy. So I use 1E-4 here. It is OK to use the IA bounds, which carries the right gradient. It would be bad if a bound could be active, but we use the MIP bound which drops the gradient. Hence a larger tolerance can avoid such problem.
